### PR TITLE
fix(subscription): close the subscription when a new schema has been detected

### LINKED ISSUE #3320: Fix hot reload with subscription
When you hot reload the router, it's keeping the old schema for opened subscriptions even if you updated the schema.

### PR DESCRIPTION
**Pull request by bnjjj**
+ The Router will close subscriptions when the hot-reload happens only if schema is different
+ The Router will send a fatal error in subscription with an error code set to `SUBSCRIPTION_SCHEMA_RELOAD`.
For example:

```json
{
  "errors": [
    {
      "message": "subscription has been closed due to a schema reload",
      "extensions": {
        "code": "SUBSCRIPTION_SCHEMA_RELOAD"
      }
    }
  ]
}
```

If a client receive that kind of error they should automatically reconnect the subscription.

Fixes #3320 

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

**Review by garypen — COMMENTED** [review]
I like the idea of a state message bus, but I think it needs more polishing and design than the approach here. For example, there are components (readiness/liveness) which are more interested in the state of the router than just observing changes to configuration/schema within the running state. They could/should probably share the same transport.

**Comment by bnjjj** [comment]
@garypen I agree with your thoughts, we could improve the way it's handle. I'm used to say, do not over engineer something if you don't need it now. As the topic of this PR is mainly to fix the current bug around hot reload with subscription and that the API is only `pub(crate)` I'm tempted to say we should try to merge this PR to fix the bug, then start to implement the readiness/liveness stuff in a follow-up PR to really find a better way to handle these cases. What do you think ?

File: apollo-router/src/state_machine.rs [hunk 1]
```diff
@@ -45,6 +48,9 @@ pub(crate) struct ListenAddresses {
     pub(crate) extra_listen_addresses: Vec<ListenAddr>,
 }
 
+#[allow(clippy::type_complexity)]
+pub(crate) static ROUTER_UPDATED: Lazy<RouterBroadcasts> = Lazy::new(RouterBroadcasts::new);
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply apollo-router/src/state_machine.rs:52]: Still not keen on this, suspect `http_server_factory` is a better place, let's talk this through.

**Review by StephenBarlow — APPROVED** [review]

**Comment by garypen** [comment]
> @garypen I agree with your thoughts, we could improve the way it's handle. I'm used to say, do not over engineer something if you don't need it now. As the topic of this PR is mainly to fix the current bug around hot reload with subscription and that the API is only `pub(crate)` I'm tempted to say we should try to merge this PR to fix the bug, then start to implement the readiness/liveness stuff in a follow-up PR to really find a better way to handle these cases. What do you think ?

I think it's ok to file an issue to introduce a "better way" and reference this and live/ready cases as the first two initial consumers.

File: apollo-router/src/notification.rs [hunk 2]
```diff
@@ -843,6 +873,36 @@ where
...
+        broadcast::Sender<Arc<Schema>>,
+        broadcast::Receiver<Arc<Schema>>,
+    ),
+}
+
+impl RouterBroadcasts {
+    pub(crate) fn new() -> Self {
+        Self {
+            configuration: broadcast::channel(2),
+            schema: broadcast::channel(2),
```
Comment (garypen) [inline apollo-router/src/notification.rs:891]: Why 2?

File: apollo-router/src/router_factory.rs [hunk 3]
```diff
@@ -541,7 +574,7 @@ pub(crate) async fn create_plugins(
 
         Err(BoxError::from(format!(
             "there were {} configuration errors",
-            errors.len()
+            dbg!(errors).len()
```
Comment (garypen) [inline apollo-router/src/router_factory.rs:577]: Is that `dbg!` desired?

**Review by garypen — APPROVED** [review]
I like the way the broadcasts are added into the existing Notify. This looks a lot cleaner than the previous iteration.

Great work!

I'll think about how we might be able to use this in the ready/live checks. 

I'm approving with a couple of small questions that your should look at before merging and ignore or action.

File: apollo-router/src/router_factory.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router/src/router_factory.rs:577]: No ! Thanks

File: apollo-router/src/notification.rs (hunk 2 above)
Comment (bnjjj) [reply apollo-router/src/notification.rs:891]: I'll switch to 1

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**